File: frontend/mobile/apple/iOS/My Home Agent/My Home Agent/Models/Media.swift
import Foundation
import UIKit

struct Media: Identifiable, Codable {
    let id: String
    let propertyRef: String?
    let serviceRequest: String?
    let serviceReport: String?
    let reportPhotoType: String?
    let uploader: String
    let file: String
    let fileType: String
    let fileSize: Int
    let title: String
    let description: String
    let uploadDate: Date
    var fileUrl: String
    private let _thumbnailSmallUrl: String?
    private let _thumbnailMediumUrl: String?
    private let _thumbnailLargeUrl: String?
    
    // Computed properties that rewrite thumbnail URLs
    var thumbnailSmallUrl: String? {
        guard let url = _thumbnailSmallUrl else { return nil }
        return NetworkManager.shared.rewriteMediaURL(url)
    }
    
    var thumbnailMediumUrl: String? {
        guard let url = _thumbnailMediumUrl else { return nil }
        return NetworkManager.shared.rewriteMediaURL(url)
    }
    
    var thumbnailLargeUrl: String? {
        guard let url = _thumbnailLargeUrl else { return nil }
        return NetworkManager.shared.rewriteMediaURL(url)
    }
    let isImage: Bool
    let isVideo: Bool
    let mediaType: MediaType
    let mediaSubType: String
    
    // Computed property to detect 3D models
    var is3DModel: Bool {
        // Check if it's a 3D_MODEL type
        if mediaType == .THREE_D_MODEL {
            return true
        }
        
        // Check mime type
        if mimeType.contains("usdz") || mimeType.contains("model/") {
            return true
        }
        
        // Fallback: Check if it's a FILE type (legacy)
        //if mediaType == .FILE && originalFilename.lowercased().hasSuffix(".usdz") {
        //    return true
        //}
        
        return false
    }
    let locationType: String
    let locationSubType: String
    let locationDisplay: String?
    let parentType: ParentType
    let originalFilename: String
    let mimeType: String
    let processingStatus: ProcessingStatus
    let isReady: Bool
    
    // UI properties
    var image: UIImage?
    
    enum CodingKeys: String, CodingKey {
        case id
        case propertyRef
        case serviceRequest
        case serviceReport
        case reportPhotoType
        case uploader
        case file
        case fileType
        case fileSize
        case title
        case description
        case uploadDate
        case fileUrl
        case _thumbnailSmallUrl = "thumbnailSmallUrl"
        case _thumbnailMediumUrl = "thumbnailMediumUrl"
        case _thumbnailLargeUrl = "thumbnailLargeUrl"
        case isImage
        case isVideo
        case mediaType
        case mediaSubType
        case locationType
        case locationSubType
        case locationDisplay
        case parentType
        case originalFilename
        case mimeType
        case processingStatus
        case isReady
    }
}

enum MediaType: String, Codable {
    case IMAGE
    case VIDEO
    case FILE
    case THREE_D_MODEL = "3D_MODEL"
    case UNKNOWN
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        self = MediaType(rawValue: rawValue) ?? .UNKNOWN
    }
}

enum ParentType: String, Codable {
    case PROPERTY
    case SERVICE_REQUEST
    case SERVICE_REPORT
    case UNKNOWN
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        self = ParentType(rawValue: rawValue) ?? .UNKNOWN
    }
}

enum ProcessingStatus: String, Codable {
    case READY
    case PENDING
    case FAILED
    case UNKNOWN
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        self = ProcessingStatus(rawValue: rawValue) ?? .UNKNOWN
    }
}

File: frontend/mobile/google/android/MyHomeAgent/app/src/main/java/com/hestami_ai/myhomeagent/data/model/Media.kt
package com.hestami_ai.myhomeagent.data.model

import com.google.gson.annotations.SerializedName
import java.util.Date

/**
 * Media type enumeration matching iOS MediaType.
 */
enum class MediaType {
    @SerializedName("IMAGE") IMAGE,
    @SerializedName("VIDEO") VIDEO,
    @SerializedName("FILE") FILE,
    @SerializedName("3D_MODEL") THREE_D_MODEL,
    UNKNOWN;

    companion object {
        @Suppress("unused")
        fun fromString(value: String): MediaType {
            return entries.find { it.name == value || 
                (it == THREE_D_MODEL && value == "3D_MODEL") 
            } ?: UNKNOWN
        }
    }
}

/**
 * Parent type enumeration matching iOS ParentType.
 */
enum class ParentType {
    @SerializedName("PROPERTY") PROPERTY,
    @SerializedName("SERVICE_REQUEST") SERVICE_REQUEST,
    @SerializedName("SERVICE_REPORT") SERVICE_REPORT,
    UNKNOWN;

    companion object {
        @Suppress("unused")
        fun fromString(value: String): ParentType {
            return entries.find { it.name == value } ?: UNKNOWN
        }
    }
}

/**
 * Processing status enumeration matching iOS ProcessingStatus.
 */
enum class ProcessingStatus {
    @SerializedName("READY") READY,
    @SerializedName("PENDING") PENDING,
    @SerializedName("FAILED") FAILED,
    UNKNOWN;

    companion object {
        @Suppress("unused")
        fun fromString(value: String): ProcessingStatus {
            return entries.find { it.name == value } ?: UNKNOWN
        }
    }
}

/**
 * Media model matching iOS Media.swift implementation.
 */
data class Media(
    @SerializedName("id")
    val id: String,

    @SerializedName("property_ref")
    val propertyRef: String? = null,

    @SerializedName("service_request")
    val serviceRequest: String? = null,

    @SerializedName("service_report")
    val serviceReport: String? = null,

    @SerializedName("report_photo_type")
    val reportPhotoType: String? = null,

    @SerializedName("uploader")
    val uploader: String,

    @SerializedName("file")
    val file: String,

    @SerializedName("file_type")
    val fileType: String,

    @SerializedName("file_size")
    val fileSize: Int,

    @SerializedName("title")
    val title: String,

    @SerializedName("description")
    val description: String,

    @SerializedName("upload_date")
    val uploadDate: Date,

    @SerializedName("file_url")
    val fileUrl: String,

    @SerializedName("thumbnail_small_url")
    val thumbnailSmallUrl: String? = null,

    @SerializedName("thumbnail_medium_url")
    val thumbnailMediumUrl: String? = null,

    @SerializedName("thumbnail_large_url")
    val thumbnailLargeUrl: String? = null,

    @SerializedName("is_image")
    val isImage: Boolean,

    @SerializedName("is_video")
    val isVideo: Boolean,

    @SerializedName("media_type")
    val mediaType: MediaType,

    @SerializedName("media_sub_type")
    val mediaSubType: String,

    @SerializedName("location_type")
    val locationType: String,

    @SerializedName("location_sub_type")
    val locationSubType: String,

    @SerializedName("location_display")
    val locationDisplay: String? = null,

    @SerializedName("parent_type")
    val parentType: ParentType,

    @SerializedName("original_filename")
    val originalFilename: String,

    @SerializedName("mime_type")
    val mimeType: String,

    @SerializedName("processing_status")
    val processingStatus: ProcessingStatus,

    @SerializedName("is_ready")
    val isReady: Boolean
) {
    /**
     * Check if this media is a 3D model.
     */
    @Suppress("unused")
    val is3DModel: Boolean
        get() = mediaType == MediaType.THREE_D_MODEL ||
                mimeType.contains("usdz") ||
                mimeType.contains("model/")
}
